src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -27,3 +36,792 @@ def deduce_line_indent(raw_segment: RawSegment, root_segment: BaseSegment) -> st
+            closing_trough = (
+                indent_points[-1].initial_indent_balance
+                + indent_points[-1].indent_trough
+            )
+        else:
+            closing_trough = (
+                indent_points[-1].initial_indent_balance
+                + indent_points[-1].indent_impulse
+            )
+
+        # On the way up we're looking for whether the ending balance
+        # was an untaken indent on the way up.
+        if (
+            # Edge case: if closing_balance > starting balance
+            # but closing_trough isn't, then we shouldn't insert
+            # a new line. That means we just dropped back down to
+            # close the untaken newline.
+            closing_trough > starting_balance
+            and closing_trough in indent_points[-1].untaken_indents
+        ):
+            # It was! Force a new indent there.
+            for ip in indent_points:
+                if ip.closing_indent_balance == closing_trough:
+                    target_point_idx = ip.idx
+                    desired_indent = single_indent * (
+                        ip.closing_indent_balance - len(ip.untaken_indents)
+                    )
+                    break
+            else:  # pragma: no cover
+                raise NotImplementedError("We should always find the relevant point.")
+            reflow_logger.debug(
+                "    Detected missing +ve line break @ line %s. Indenting to %r",
+                elements[target_point_idx + 1].segments[0].pos_marker.working_line_no,
+                desired_indent,
+            )
+            target_point = cast(ReflowPoint, elements[target_point_idx])
+            new_fixes, new_point = target_point.indent_to(
+                desired_indent, before=elements[target_point_idx + 1].segments[0]
+            )
+            elements[target_point_idx] = new_point
+            fixes += new_fixes
+            # Keep track of the indent we forced
+            forced_indents.append(closing_balance)
+
+    # Or the way down.
+    elif closing_balance < starting_balance:
+        # On the way down we're looking for indents which *were* taken on
+        # The way up, but currently aren't on the way down. We slice so
+        # that the _last_ point isn't evaluated, because that's fine.
+        for ip in indent_points[:-1]:
+            # Is line break, or positive indent?
+            if ip.is_line_break or ip.indent_impulse >= 0:
+                continue
+            # It's negative, is it untaken?
+            if (
+                ip.initial_indent_balance in ip.untaken_indents
+                and ip.initial_indent_balance not in forced_indents
+            ):
+                # Yep, untaken.
+                continue
+            # Edge Case: Semicolons. For now, semicolon placement is a little
+            # more complicated than what we do here. For now we don't (by
+            # default) introduce missing -ve indents before semicolons.
```

Comment:
What's `-ve`?

Reply:
Ah - that's an abbreviation of _negative_. i.e. a point where the indent balance decreases.